lib/interpreters/periodic_visits.rb

```diff
@@ -371,36 +373,63 @@ def associate_skills(new_vehicle, vehicle_day_index)
       end
     end
 
-    def compute_possible_days(vrp)
-      # for each of this service's visits, computes first and last possible day to be assigned
-      vrp.services.each{ |service|
-        day = @schedule_start
-        nb_services_seen = 0
-
-        # first possible day
-        while day <= @schedule_end && nb_services_seen < service.visits_number
-          if service.unavailable_days.include?(day) || vrp.vehicles.none?{ |v| v.available_at(day) }
-            day += 1
+    def possible_days_from_possible_day(service, possible_days, vehicles, lapse, options)
+      day = @schedule_start
+      nb_services_seen = 0
+      set = []
+      while day <= @schedule_end && nb_services_seen < service.visits_number
+        # it is possible that user provided last_performed_visit_date/day_index or filled first_possible_days field
+        # we assume if there are X values in first_possible_days then it corresponds to X first visits of this service
+        # user needs to provide nil value if he wants to skip one service when providing first_possible_days
+        day =
+          if lapse
+            [possible_days[nb_services_seen], day].compact.max
           else
-            service.first_possible_days += [day]
-            nb_services_seen += 1
-            day += service.minimum_lapse || 1
+            # case with maximum_lapse, but maximum_lapse is nil
+            @schedule_end - (service.visits_number - nb_services_seen - 1) * (service.minimum_lapse || 1)
           end
+
+        while (service.unavailable_days.include?(day) || vehicles.none?{ |v| v.available_at(day) } || # day is not available
+              nb_services_seen > 0 && day <= set[nb_services_seen - 1] ||
+              options[:first_possible_days][nb_services_seen] && options[:first_possible_days][nb_services_seen] > day) &&
+              day.between?(@schedule_start, @schedule_end)
+          day += options[:is_min_lapse] ? 1 : -1
         end
 
-        # last possible day
-        day = @schedule_end
-        nb_services_seen = 0
-        while day >= @schedule_start && nb_services_seen < service.visits_number
-          if service.unavailable_days.include?(day) || vrp.vehicles.none?{ |v| v.available_at(day) }
-            day -= 1
-          else
-            service.last_possible_days += [day]
+        return set unless day.between?(@schedule_start, @schedule_end)
+
+        set << day
+        nb_services_seen += 1
```

Comment:
```suggestion
        visit_index += 1
```

Reply:
If we are looking after every visit_index, wouldn't it be easier to have (1..service.visits_number).each{ |visit_index| ... } ?

Reply:
I guess that would be another way to do it but keeping track of seen_visits would be harder.